First version of REST calls for NSC plugin

sudhirappaji opened the pull request:
This PR contains the following:
1. Rest Framework Infrastructure.
2. REST Implementation for NetworkElementEntity.
3. REST implementation for InspectionPortEntity.
4. REST implementation for InspectionHookEntity.

sudhirappaji committed:
First version of REST calls for NSC plugin

emanoelxavier commented:
@sudhirappaji I have not looked at this PR yet but I believe you are currently working on https://github.com/opensecuritycontroller/security-mgr-sample-plugin/pull/33 and therefore not blocked on this one but please let me know otherwise.

sudhirappaji commented:
@emanoelxavier , I have just released the security-mgr-sample-plugin for review for device updates and will continue to work on testing this changes on deployed OSC. so not blocked for time being and I assume that this nsc will not be part of this release?

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+@Path("/nw")
+@Consumes({ MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML })
+@Produces({ MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML })
+public class InspectionApis {
```
I think you need to have three different APIs here:
1. networkelementapis mapped to /networkelements
2. inspectionport mapped to /inspectionports
3. inspectionhook mapped to /inspectionhooks
This will result in three different classes. @arvindn05 do you agree?

sudhirappaji replied:
ok will take care

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+
+    }
+
+    @Path("/ipAddr/{ipAddr}/netelem")
```
ipAddr -> controller/{id} . I understand we are using the ip address to identify. However we should keep the API more generic, the ip here is just being used as an id for the controller.

sudhirappaji replied:
ok, will take care.

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+
+        init(ipAddr);
+        return this.txControl.required(() -> {
+            NetworkElementEntity element = this.utils.txNetworkElementEntityByElementId(entity.getElementId());
```
Typically the id wouldnt be provided if you are doing a create so instead if the user is passing an Id in the request/entity you should probably throug an illegalargumentexception. You may repeat the same pattern for the other apis

sudhirappaji replied:
If I look at the NetworkElementEntity, id value is not generated automatically as annotation generated value is not used. so user passing the value is valid here.
@dmitryintel , please comment.

emanoelxavier commented:
@sudhirappaji I havent fully finished the review yet but I added a few comments. Thx.

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+            if (element == null) {
+                LOG.info(String.format(
+                        "Attempt to delete network element for id %s and network element not found, no-op.",
+                        "" + elementId));
```
" not needed

sudhirappaji replied:
ok

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+        init(ipAddr);
+
+        return this.txControl.required(() -> {
+            NetworkElementEntity element = this.utils.txNetworkElementEntityByElementId(entity.getElementId());
```
This should be the url elementid, you can add an input validation:
```if (ent.geteleid() != null && ent.geteleid() != elementid) throw....```

sudhirappaji replied:
I will skip this validation and take only the url elementid.

sudhirappaji replied:
I would like to avoid this checking and use url elementId directly.

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+    @DELETE
+    public void deleteNetworkElement(@PathParam("ipAddr") String ipAddr, @PathParam("elementId") String elementId)
+            throws Exception {
+        LOG.info(String.format("Deleting the network element for id %s ", "" + elementId));
```
"" not needed. Also look at other places

sudhirappaji replied:
ok

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+        return this.txControl.supports(() -> {
+            List<? extends InspectionPortEntity> elements = this.utils.txInspectionPortEntities();
+            List<String> elementList = new ArrayList<String>();
+            if (elements.isEmpty()) {
```
revert the if to eliminate one of the returns

sudhirappaji replied:
ok

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+
+    @Path("/ipAddr/{ipAddr}/insport/{insportId}")
+    @GET
+    public InspectionPortEntity getInspectionPortElement(@PathParam("ipAddr") String ipAddr,
```
getInspectionPortElement -> getInspectionPort

sudhirappaji replied:
ok

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+        return (InspectionPortEntity) this.api.getInspectionPort(inspectionPort);
+    }
+
+    @Path("/ipAddr/{ipAddr}/insport/{insportId}/inshook")
```
I dont think you need to make the insphook a child of the insp port. There is a many to one relationship but I dont believe it must be a parent.

sudhirappaji replied:
ok

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+    @Path("/ipAddr/{ipAddr}/insport/{insportId}/inshook")
+    @POST
+    public String createInspectionHook(@PathParam("ipAddr") String ipAddr, @PathParam("insportId") String insportId,
+            InspectionHookEntity entity)
```
This entity comes with other entity fields in it, all of which implement an SDK *Element interface. Given that, can you simply call the api installInspectionHook passing those fields? That method already does a good part of what you are doing here.

sudhirappaji replied:
ok

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+    @PUT
+    public InspectionHookEntity updateInspectionHook(@PathParam("ipAddr") String ipAddr,
+            @PathParam("insportId") String insportId, @PathParam("insportId") String inshookId,
+            InspectionHookEntity entity) throws Exception {
```
Same here, can you simply call the updateInspectionhook from the SDK implementation?

sudhirappaji replied:
ok

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionApis.java:
```diff
@@ -0,0 +1,403 @@
+        return this.txControl.supports(() -> {
+            List<InspectionHookEntity> elements = this.utils.txInspectionHookEntities(insportId);
+            List<String> elementList = new ArrayList<String>();
+            if (elements.isEmpty()) {
```
revert if

sudhirappaji replied:
ok

emanoelxavier commented on src/main/java/org/osc/controller/nsc/utils/RedirectionApiUtils.java:
```diff
@@ -201,10 +201,40 @@ public InspectionPortEntity txInspectionPortEntityById(String id) {
         return this.em.find(InspectionPortEntity.class, id);
     }
 
+    public List<InspectionHookEntity> txInspectionHookEntities(String inspectionPort) {
```
I dont think this will be necessary if you dont make the inspectionport the parent.

sudhirappaji replied:
ok

sudhirappaji committed:
Addressed the review comments.

sudhirappaji committed:
Merge branch 'master' into nsc_stubs

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionHookApis.java:
```diff
@@ -0,0 +1,163 @@
+        LOG.info("Listing inspection hook ids'");
+        init(id);
+
+        return this.txControl.supports(() -> {
```
instead of having this txcontrol here you can actually create a method in the `SampleSdnRedirectionApi` and call it here. This should be fine since this class has the type `SampleSdnRedirectionApi` instead of the SDK interface. This way the `map` approach I mentioned above can still work here since you would no longer need the txControl.

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionPortApis.java:
```diff
@@ -0,0 +1,177 @@
+    public InspectionPortEntity updateInspectionPort(@PathParam("id") String id,
+            @PathParam("portId") String portId, InspectionPortEntity entity) throws Exception {
+
+        LOG.info(String.format("Updating the inspection port element id %s ", portId));
```
Similar comment here, you could also move this code as a method into `SampleSdnRedirectionApi`

sudhirappaji commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionHookApis.java:
```diff
@@ -0,0 +1,163 @@
+
+    }
+
+    @Path("/controller/{id}/inshook")
```
@arvindn05 , could you please feedback?

sudhirappaji replied:
@arvindn05 ,I am falling short of time and need to complete this PR. I shall go ahead and implement the same. Hope it is fine with you.

sudhirappaji commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionHookApis.java:
```diff
@@ -0,0 +1,163 @@
+            throw new IllegalArgumentException("Attempt to delete null inspection hook port");
+        }
+        LOG.info(String.format("Deleting the inspection hook element for id %s ", inshookId));
+        init(id);
```
@arvindn05 , could you please feedback?

sudhirappaji replied:
@arvindn05 , @arvindn05 , I am running out of time to complete this PR and will go ahead and implement the same. Hope it is fine with you.

sudhirappaji replied:
@arvindn05 ,I am falling short of time and need to complete this PR. I shall go ahead and implement the same. Hope it is fine with you.

sudhirappaji committed:
Merge branch 'master' into nsc_stubs

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/SampleSdnServletDelegate.java:
```diff
@@ -0,0 +1,90 @@
+
+@Component(name = "nsc.servlet", service = Servlet.class, property = {
+
+        HTTP_WHITEBOARD_SERVLET_NAME + "=" + "OSC-API", HTTP_WHITEBOARD_SERVLET_PATTERN + "=/testplugin/sdn/*",
```
Can the URL be just nsc/*?

Also the Servlet name should reflect the nsc as well. Similar comment for the security manager api.

OSC-API should only be used for the core api. Whiteboard target i beleive needs to be OSC-API.

Correct me if i am wrong

sudhirappaji replied:
ok, will change the URL to nsc/* and also Servlet name to "NSC-API". I will adapt this as well for securitysample plugin.

emanoelxavier replied:
@arvindn05 the reason for having `testplugin` is to explicitly call out in the url this is a test entry point. Having simply `nsc` hides that. Also the plugin is really not nsc, but really a sample so I recommend not having nsc in the url name. For the manager the url is `/sm/` so what would the name be for consistency? ism?  The urls `/testplugin/sdn` and `/testplugin/sm` seem clear and consistent to me.  In the future we can also have /testplugin/sdn-sfc, for instance. @sudhirappaji  before you change this please share what will be the name in the url for both manager and sdn.

arvindn05 replied:
makes sense. But instead of test plugin can we use consistent terminology maybe 

>sample/mgr/ism/*
>sample/sdn/nsc/*

emanoelxavier replied:
@sudhirappaji @arvindn05 the names above are good we can go with that.

sudhirappaji replied:
ok, will handle as proposed.

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionHookApis.java:
```diff
@@ -0,0 +1,163 @@
+    private EntityManager em;
+    private RedirectionApiUtils utils;
+    private SampleSdnRedirectionApi sdnApi;
+    @Reference(target = "(osgi.local.enabled=true)")
```
add line spaces in between the declaration blocks.

I see 3 blocks , the log, the private variable and private @reference variables.

also add lines spaces between each of the annotated variables similiar to the sytle in OSC. I will try to enforce this via checkstyle but for now please add line space like below

```
@Reference(target = "(osgi.unit.name=nsc-mgr)")
private EntityManagerFactoryBuilder builder;

@Reference(target = "(osgi.jdbc.driver.class=org.h2.Driver)")
private DataSourceFactory jdbcFactory;
```

sudhirappaji committed:
Addressed review comments.

sudhirappaji commented:
@arvindn05  @emanoelxavier , I have addressed all of the review remarks.

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/NetworkElementApis.java:
```diff
@@ -0,0 +1,133 @@
+    @Reference(target = "(osgi.local.enabled=true)")
+    private JPAEntityManagerProviderFactory resourceFactory;
+
+    @Path("/{elementId}")
```
no need to add the id on create APIs. Same thing for create inspection hook and inspection ports. I dont think we use the entity id on the creation apis of the security manager sample plugin.

sudhirappaji replied:
@emanoelxavier , Id is used for creation of NetworkElementEntity as we discussed during one of the remarks earlier. For InspectionPort and InspectionHook, Id is not passed as it is auto generated and similar to as in security sample code. 
This is how it was designed for the networkelemententity.
@dmitryintel , please feedback why Id is not auto generated for networkelemententity? did you encounter any issue when auto generation is enabled?

emanoelxavier replied:
got it

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionPortApis.java:
```diff
@@ -0,0 +1,177 @@
+
+    }
+
+    @Path("/controller/{id}/insport")
```
inspectionPorts . Please review all the APIs and ensure the names/paths are consistent.

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionHookApis.java:
```diff
@@ -0,0 +1,135 @@
+                entity.getEncType(), entity.getOrder(), entity.getFailurePolicyType());
+    }
+
+    @Path("/{inshookId}")
```
camel case => inshookId -> insHookId or InspectionHookId

emanoelxavier commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionPortApis.java:
```diff
@@ -0,0 +1,136 @@
+        return inspectionPort == null ? null : inspectionPort.getElementId();
+    }
+
+    @Path("/{portId}")
```
inspectionPortId

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -79,6 +77,18 @@ public InspectionHookElement getInspectionHook(NetworkElement inspectedPort, Ins
         }
     }
 
+    public List<String> getInspectionHooks() throws Exception {
```
lets name these methods more distinctly. getInspectionHooksIds()

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -269,6 +263,18 @@ public InspectionPortElement getInspectionPort(InspectionPortElement inspectionP
         return ipEntity;
     }
 
+    public List<String> getInspectionPorts() throws Exception {
```
same comment as above.

Also please group these  helper methods and move them to a section below all the interface methods and name the block something like 

// Support methods for API

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -411,6 +427,23 @@ public NetworkElement updateNetworkElement(NetworkElement portGroup, List<Networ
         });
     }
 
+    public NetworkElement updateNetworkElement(NetworkElement nwElement) throws Exception {
```
please name these methods distinctly as well. This is basically creating a port even though internally we are creating a Network Element. So basically this method could be createPort, updatePort etc

In NSC context, *NetworkElement methods refer to port group(in neutron sfc they refer to PPG's). Adding yet another overload on top of *NetworkElement method seems confusing.

sudhirappaji replied:
@arvindn05 , I see an issue because in the sdn-controller-api sdk, I already see API's like getNetworkElements, etc. So renaming just the support methods like createPorts, getPorts etc will lead to confusion. So i plan to keep it as is and refine the methods names where needed (like some comments that is applicable).

arvindn05 replied:
its much harder to see the distinction between the support methods and the sdk methods.

updateNetworkElement(2  params) (SDK method) -> updateNetworkElement (single param)

work completely differently. SDK referes to updating port group and updateNetworkElement refers to updating a port.

@emanoelxavier your thougths?

emanoelxavier replied:
Currently the consistency pattern is entity name (networkelemententity) -> rest api name (network element resource). The question here is about how to name the helper methods in between.  I like the idea of calling it `port` however if we are doing that then the entity should be `portentity` and the rest resource `port` instead of `network element`. In this case the helper methods then could be called `port`. This would make it all consistent across the board. But this is also a bigger change since we would rename the tables and entities as well. My recommendation would be rename the helper methods as arvind mention to indicate that intent (these things are really ports) and place a TODO on the entity and rest api to have it renamed to port as well in the future. Makes sense?

sudhirappaji replied:
@emanoelxavier, I will implement this way and put todo for the entity and the rest api's.

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/api/RestServerApiUtils.java:
```diff
@@ -0,0 +1,72 @@
+        return sdnApiHashMap.get(id);
+    }
+
+    public static EntityManager createEntityHandle(JPAEntityManagerProviderFactory resourceFactory,
```
why do we need to duplicate this code from SampleSdnControllerApi?

can we just inject the SampleSdnControllerApi into the rest api servlet?

sudhirappaji replied:
@arvindn05 , the possibility i see is to have a common method and this method could be called from SampleSdnControllerApi and by the rest layer. This method could be defined in redirectionutil class. Let me know what you think or any other possibility you see? and will wait for your feedback before any change on this front.

sudhirappaji replied:
@arvindn05 , I am waiting for your feedback for submitting this PR.

arvindn05 replied:
@sudhirappaji can you reply to my original question? Why cant we inject the SampleSdnControllerApi?

sudhirappaji replied:
@arvindn05, yes it should be feasible to inject using Inject annotation. Will confirm after checking it. But need to implement VC passed as a parameter to create method.

sudhirappaji replied:
@arvindn05, I tried injecting SampleSdnControllerApi in the rest layer. This always results in null exception/ as injection is not working. I tried to check/investigate the possible causes, unfortunately did not find any solution. Please guide how to go about? I already spent couple of hours on this topic, but no solution in sight. my suggestion is I can put this as TODO item.

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -87,11 +97,8 @@ public String installInspectionHook(NetworkElement inspectedPort, InspectionPort
         if (inspectedPort == null) {
             throw new IllegalArgumentException("Attempt to install an Inspection Hook with no Inspected Port");
         }
-        if (inspectionPort == null) {
```
why are we removing all these checks?

sudhirappaji replied:
@arvindn, oh it is by accident and will fix it

sudhirappaji committed:
Rework based on the review comments.

sudhirappaji commented:
@arvindn05, @emanoelxavier - Addressed the review comments, please check. I will meanwhile update the branch with the master as part of next iteration.

arvindn05 commented on pom.xml:
```diff
@@ -39,6 +39,75 @@
+			<version>6.0.0</version>
+			<scope>provided</scope>
+		</dependency>
+		<dependency>
```
please add right scopes for each of the dependencies.

same applies to manager plugin...maybe create an issue on the sample plugin so it can be tracked and fixed

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -495,4 +500,135 @@ public void close() throws Exception {
+        return this.txControl.supports(() -> {
+
+            List<InspectionHookEntity> elements = this.utils.txInspectionHookEntities();
+            List<String> elementList = new ArrayList<String>();
```
you dont need the element list or the check for isEmpty

you could just do this.utils.txInspectionHookEntities().stream().map(InspectionHookEntity::getHookId).collect(Collectors.toList())

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -495,4 +500,135 @@ public void close() throws Exception {
+    public Element updateInspectionPort(InspectionPortElement inspectionPort) throws Exception {
+
+        InspectionPortEntity inspectionPortEntity = (InspectionPortEntity) getInspectionPort(inspectionPort);
+        if (inspectionPortEntity == null) {
```
use org.osc.controller.nsc.utils.RedirectionApiUtils.throwExceptionIfNullElement(InspectionPortElement)

sudhirappaji replied:
@arvindn05 , these are private methods and will make it public.

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -495,4 +500,135 @@ public void close() throws Exception {
+        }
+
+        return this.txControl.required(() -> {
+            inspectionPortEntity.setEgressPort((NetworkElementEntity) inspectionPort.getEgressPort());
```
where do we validate the egress and ingress ports exist?

Since we have unit tests for nsc plugin...please use the tests to validate all these scenarios

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -495,4 +500,135 @@ public void close() throws Exception {
+            List<? extends InspectionPortEntity> elements = this.utils.txInspectionPortEntities();
+            List<String> elementList = new ArrayList<String>();
+
+            if (!elements.isEmpty()) {
```
same as above

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -495,4 +500,135 @@ public void close() throws Exception {
+        return this.txControl.required(() -> {
+            element.setMacAddresses(nwElement.getMacAddresses());
+            element.setPortIPs(nwElement.getPortIPs());
+            element.setDeviceOwnerId(((NetworkElementEntity) nwElement).getDeviceOwnerId());
```
this is an unsafe cast. How are we guaranteed to receive NetworkElementEntity? maybe just modify the method to take a NetworkElementEntity

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -495,4 +500,135 @@ public void close() throws Exception {
+            NetworkElementEntity element = this.utils.txNetworkElementEntityByElementId(id);
+
+            if (element == null) {
+                LOG.warn(String.format(
```
with slf4j logging no longer needs to use string.format
Use below if we are just logging the message and not wrapping it with an exception

`LOG.warn("Attempt to delete network element for id {} and network element not found, no-op.", id);`

arvindn05 commented on src/main/java/org/osc/controller/nsc/model/VirtualizationConnectorElementImpl.java:
```diff
@@ -0,0 +1,109 @@
+
+import org.osc.sdk.controller.element.VirtualizationConnectorElement;
+
+public class VirtualizationConnectorElementImpl implements VirtualizationConnectorElement {
```
add javadoc to indicate lot of the properties are not returned.

throw UnsupportedOperationException for all methods which are not implmented and returning null etc

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionHookApis.java:
```diff
@@ -0,0 +1,125 @@
+import org.slf4j.LoggerFactory;
+
+@Component(service = InspectionHookApis.class)
+@Path("/controller/{id}/inspectionHooks")
```
rename id to controller id so its clear its not inspectionhook id.

same comment for the id variables in the methods

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionHookApis.java:
```diff
@@ -0,0 +1,125 @@
+        LOG.info(String.format("Creating inspection hook with inspection port id %s",
+                entity.getInspectionPort().getElementId()));
+
+        return RestServerApiUtils.getSdnRedirectionApi(id, this.api).installInspectionHook(entity.getInspectedPort(),
```
i would just use the api directly. Eliminate the use of RestServerApiUtils. 
The API lifecycle/caching is managed by OSGI already. we should not duplicate that functionality.

sudhirappaji replied:
@arvindn05 , I have eliminated getSdnRedirectionApi from RestServerApiUtils and directly using the api's in the method. But RestServerApiUtils I would still like to retain, as the HASH updates between Port, inspectionPort and inspectionHook are managed in this utility along with the id validatin method.

arvindn05 replied:
not sure i understand this comment. i dont see any id validation either.

It just seems like you are caching the sdn redirection api in a static map.
if you are caching the api's, who is clearing the map? How do we know the api in the map is still valid? the DB connection can close underneath...

sudhirappaji replied:
@arvindn05,yes each api hold entity manager handle, which is used to access the db. The cache entity is not cleared, good point. I am thinking of flushing this cache entries as part of 'close' method of networkapis or other option is when the Ser let is getting destroyed or any other option?
Regarding Id validation it can be any ip address and I can check for null value as part of validation.

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionHookApis.java:
```diff
@@ -0,0 +1,125 @@
+                entity.getFailurePolicyType());
+    }
+
+    @Path("/{InspectionHookId}")
```
naming convention should be camel case.

sudhirappaji replied:
@arvindn05, this was the camel case naming proposed by Emanoel - InspectionHookId. I will change this to lower camelcase, as it is a variable to inspectionHookId, ok?

sudhirappaji replied:
@arvindn05, this was the camel case naming proposed by Emanoel - InspectionHookId. I will change this to lower camelcase, as it is a variable to inspectionHookId, ok?

arvindn05 replied:
there are 2 camel case conventions, UpperCamelCase(used here) and lowerCamelCase. 
lowerCamelCase should be used for all variable names.
https://www.thoughtco.com/using-java-naming-conventions-2034199

emanoelxavier replied:
camel case for variable ALWAYS start with a lower case, my comment had a typo "camel case => inshookId -> insHookId or InspectionHookId"

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionHookApis.java:
```diff
@@ -0,0 +1,125 @@
+    @PUT
+    public InspectionHookEntity updateInspectionHook(@PathParam("id") String id,
+            @PathParam("InspectionHookId") String InspectionHookId, InspectionHookEntity entity) throws Exception {
+        entity.setHookId(InspectionHookId);
```
do we want to validate any mismatch between URL id and the id within the object?

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/api/InspectionHookApis.java:
```diff
@@ -0,0 +1,125 @@
+    @DELETE
+    public void deleteInspectionHook(@PathParam("id") String id, @PathParam("InspectionHookId") String InspectionHookId)
+            throws Exception {
+        if (InspectionHookId == null) {
```
we should add validation only which does not exist in the api already. For example the url id and object id mismatch etc

For everything else reuse existing validation in the API implementation.

In this case the org.osc.controller.nsc.api.SampleSdnRedirectionApi.removeInspectionHook(String) already handles null id, so we dont need this check. 

Please fix other places as well.

Alternative is if we want the api to fail for null id's then make the call the api, check the payload is null and throw the exception at that point

sudhirappaji committed:
Merge branch 'master' into nsc_stubs

sudhirappaji committed:
Addressed the review comments and in addition added test cases for
the support methods in sdndirectionapi.

sudhirappaji commented:
@arvindn05 , I have made the changes as requested, please review.

arvindn05 commented on pom.xml:
```diff
@@ -39,6 +39,79 @@
+		</dependency>
+		<dependency>
+			<groupId>org.slf4j</groupId>
+			<artifactId>slf4j-log4j12</artifactId>
```
we should be using logback not log4j

sudhirappaji replied:
@arvindn05, I see logback was already there and as part of the merge this got included and will clean this up.

arvindn05 commented on pom.xml:
```diff
@@ -159,6 +231,7 @@
             <groupId>org.hibernate</groupId>
```
is hibernate really a compile time dependency? JPA should be our compile dependecy.

please take a deep look at all the dependencies as it is difficult to maintain them after the fact we introduced with wrong scope.

sudhirappaji replied:
@arvindn05, This was already there and I have not introduced this dependency. I also had a closer look at this and see that it is indeed compile time dependency as it is used in NetworkElementEntity, as it uses the following: org.hibernate.annotations.Fetch;org.hibernate.annotations.FetchMode; and also used in InspectionPortEntity: org.hibernate.annotations.GenericGenerator;

sudhirappaji replied:
@arvindn05, I had the closer look at the dependency I had put and see the following changes: javax.servlet-api, scope shall be changed to provided, jersey-server: scope shall be changed to provided,jersey-container-servlet-core: scope shall be changed provided,javax.ws.rs-api ; scope shall be changed to provided, jackson-annotations to 'runtime',jackson-core to runtime,jackson-databind to runtime,jackson-jaxrs-base to runtime, org.apache.commons.lang to provided.

arvindn05 commented on pom.xml:
```diff
@@ -99,6 +167,7 @@
             <groupId>org.osgi</groupId>
             <artifactId>org.osgi.core</artifactId>
             <version>6.0.0</version>
+            <scope>compile</scope>
```
same here...should be be provided?

sudhirappaji replied:
@arvindn05, this was already compile scope in the earlier code and shall change to provided.

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -72,8 +73,8 @@ public InspectionHookElement getInspectionHook(NetworkElement inspectedPort, Ins
         } catch (Exception e) {
-            LOG.error(String.format("Exception finding Network Element (Inspected %s ; Inspection Port %s):",
-                    "" + inspectedPort, "" + inspectionPort), e);
+            LOG.error("Exception finding Network Element (Inspected %s ; Inspection Port %s):", "" + inspectedPort,
```
%s should be replaced  by {}. see previous comment. also have string concatenation with "" + inspectedPort which is not needed

`LOG.warn("Attempt to delete network element for id {} and network element not found, no-op.", id);
`

sudhirappaji replied:
Double quote was already part of earlier code, will clean this up as proposed.

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -488,4 +494,124 @@ public void close() throws Exception {
+
+    public Element updateInspectionPort(InspectionPortElement inspectionPort) throws Exception {
+
+        InspectionPortEntity inspectionPortEntity = (InspectionPortEntity) getInspectionPort(inspectionPort);
```
lets do prevalidation of inspection port, ingrees/egress ports(both object and ids) before we get the entity out of the db

sudhirappaji replied:
@arvindn05, prevalidation of inspectionPort is already part of getInspectionPort method and hence it is not done.

sudhirappaji replied:
inspectionPort is already validated in getInspectionPort Method and for Ingress/Egress ports check is added to validate for id.

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -488,4 +494,124 @@ public void close() throws Exception {
+            this.utils.throwExceptionIfNullElement(inspectionPort.getEgressPort(), "Egress port is null");
+        }
+
+        if (inspectionPort.getEgressPort().getElementId() == inspectionPort.getIngressPort().getElementId()) {
```
use .equals comparison.

also load the ingress element from the db, and check if the egress element id provided is the same as the ingress entity id. this will eliminate a bit of code

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -488,4 +494,124 @@ public void close() throws Exception {
+        }
+
+        return this.txControl.required(() -> {
+            inspectionPortEntity.setEgressPort((NetworkElementEntity) inspectionPort.getEgressPort());
```
how is this working? inspectionport is passed in from external, we cannot make an assumption that its a NetworkElementEntity and its in the current session

sudhirappaji replied:
@arvindn05, i am checking for validity of Egress, Ingress port elements which is embedded as a part of inspection port. As I see from definition ingress is a network element type. So checking whether this Id already exists in db. Is this incorrect? Let's follow up offline.

arvindn05 replied:
how is the `inspectionPort.getEgressPort()` guaranteed to be a entity. From the signature the egrees port is only going to be a NetworkElement so its an usafe cast.

when you set the `inspectionPortEntity.setEgressPort` the port should be guaranteed to be a    NetworkElementEntity AND it has be attached to a session(using merge or some other mechanism)

As this method updateInspectionHook is a public method it can be called from anywhere and i could pass in just any network element and this would cause failure as part of the cast...

sudhirappaji replied:
Thanks I got it

arvindn05 reviewed (changes requested):
i am not sure how this works when tested...seems like lots of issue which would be caught with testing.

please setup a call to discuss.

sudhirappaji committed:
Merge  'master' into nsc_stubs

sudhirappaji committed:
Addressed the pom dependency scope and removed hash for caching the SdnRedirectionApi.

sudhirappaji commented:
@arvindn05 , Please review and I have incorporated as per your suggestion. I have removed the test cases added by Praveen, as I see test structure has changed during the merge.
@pravinrajr9 , please add the test case as per the new structure as part of this PR.

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -488,4 +494,118 @@ public void close() throws Exception {
+        InspectionPortEntity inspectionPortEntity = (InspectionPortEntity) getInspectionPort(inspectionPort);
+        this.utils.throwExceptionIfNullElement(inspectionPortEntity);
+
+        if (inspectionPort.getIngressPort().equals(null)
```
please add method like we did in the neutron sfc plugin 
org.osc.controller.nsfc.utils.RedirectionApiUtils.throwExceptionIfNullElementAndId(Element, String)

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/api/NetworkElementApis.java:
```diff
@@ -0,0 +1,154 @@
+        SampleSdnRedirectionApi sdnApi = ((SampleSdnRedirectionApi) this.api
+                .createRedirectionApi(new VirtualizationConnectorElementImpl("Sample", controllerId), "TEST"));
+
+        RedirectionApiUtils utilsApi = new RedirectionApiUtils(null, null);
```
we are not properly initializing the utils class. Either need to move the methods which do not require em and txcontrol out of this util or we need to pass in the correct em/txcontrol

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/api/NetworkElementApis.java:
```diff
@@ -0,0 +1,154 @@
+    @POST
+    public String createNetworkElement(@PathParam("controllerId") String controllerId,
+            @PathParam("elementId") String elementId, NetworkElementEntity entity) throws Exception {
+        // TODO:SUDHIR - Rename Rest API's to Port instead of NetworkElement, for eg: createPort etc.
```
should be straight forward to fix this right? Why do we want to track a TODO?

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -488,4 +494,118 @@ public void close() throws Exception {
+        }
+
+        return this.txControl.required(() -> {
+            NetworkElementEntity portElement = new NetworkElementEntity();
```
pretty much seems like a copy constructor is needed for NetworkElementEntity?

sudhirappaji committed:
Merge branch master into nsc_stubs

sudhirappaji committed:
Added Jersey unit test for REST API's and also fixed review feedback.

sudhirappaji commented:
@arvindn05 , I have added the tests needed, please review.

arvindn05 commented on src/main/java/org/osc/controller/nsc/api/SampleSdnRedirectionApi.java:
```diff
@@ -488,4 +494,124 @@ public void close() throws Exception {
+
+        PortEntity element = this.utils.findPortEntityByElementId(portElement.getElementId());
+        if (element != null) {
+            String msg = String.format("Network element already exists id: %s\n", portElement.getElementId());
```
Port element?

arvindn05 commented on src/main/java/org/osc/controller/nsc/entities/PortEntity.java:
```diff
@@ -83,6 +83,14 @@ public NetworkElementEntity(String elementId, List<String> macAddressEntities, L
+    public PortEntity(PortEntity element){
+        this.elementId = element.elementId;
+        this.deviceOwnerId = element.deviceOwnerId;
+        this.parentId = element.parentId;
```
if parentid is not applicable, can we just remove it?

sudhirappaji replied:
parentId access methods are needed, as it is defined in the Base Class.

arvindn05 commented on src/main/java/org/osc/controller/nsc/restserver/api/PortApis.java:
```diff
@@ -0,0 +1,131 @@
+        SampleSdnRedirectionApi sdnApi = ((SampleSdnRedirectionApi) this.api
+                .createRedirectionApi(new VirtualizationConnectorElementImpl("Sample", controllerId), "TEST"));
+
+        sdnApi.throwExceptionIfNullId(controllerId);
```
update createRedirectionApi to throw error in case of null id since that needs to be validated anyway. You can avoid this check for every API in that case

arvindn05 commented:
please address minor comments. @larkinscarvalho please review the unit tests.

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/InspectionHookApisTest.java:
```diff
@@ -0,0 +1,267 @@
+        return port;
+    }
+
+    private InspectionHookEntity getInspectionHookEntity() {
```
createInspectionHookEntity

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/InspectionHookApisTest.java:
```diff
@@ -0,0 +1,267 @@
+        }
+    }
+
+    private PortEntity getPortEntity() {
```
createPortEntity

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/InspectionHookApisTest.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    @Test
+    public void testDeleteInspectionHook_expectNoContent() {
```
to be consistent with other project test names

testExecute_Action_ExpectedResult -> testExecute_WithDeleteInspectionHook_ExpectNoContent

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/BaseJerseyTest.java:
```diff
@@ -0,0 +1,53 @@
+
+public class BaseJerseyTest extends JerseyTest {
+
+    protected final String authorizationHeader = "Authorization";
```
Final variable uppercase

AUTHORIZATION_HEADER

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/InspectionPortApisTest.java:
```diff
@@ -0,0 +1,277 @@
+    }
+
+    @Test
+    public void testGetInspectionPort_expectStatusOk() {
```
please update the test method names

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/InspectionPortApisTest.java:
```diff
@@ -0,0 +1,277 @@
+        }
+    }
+
+    private PortEntity getPortEntity() {
```
createPortEntity

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/InspectionPortApisTest.java:
```diff
@@ -0,0 +1,277 @@
+        return port;
+    }
+
+    private InspectionPortEntity getInspectionPortEntity() {
```
createInspectionPortEntity

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/PortApisTest.java:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    @Test
+    public void testGetPortIds_expectStatusOk() {
```
please update the test method names

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/RedirectionApiUtilsTest.java:
```diff
@@ -77,8 +77,8 @@ public void testUtils_PersistHookOnly_InspectedNetworkElementByElementIdFound()
         // Act.
-        NetworkElementEntity foundNE = this.txControl.required(() -> {
-            NetworkElementEntity e = this.utils.txNetworkElementEntityByElementId(inspected.getElementId());
+        PortEntity foundNE = this.txControl.required(() -> {
```
rename foundNE as you have refactored NetworkElementEntity -> PortEntity

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/RedirectionApiUtilsTest.java:
```diff
@@ -77,8 +77,8 @@ public void testUtils_PersistHookOnly_InspectedNetworkElementByElementIdFound()
-        NetworkElementEntity foundNE = this.txControl.required(() -> {
-            NetworkElementEntity e = this.utils.txNetworkElementEntityByElementId(inspected.getElementId());
+        PortEntity foundNE = this.txControl.required(() -> {
+            PortEntity e = this.utils.txPortEntityByElementId(inspected.getElementId());
```
valid variable name

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/SampleSdnRedirectionApiTest.java:
```diff
@@ -523,4 +524,214 @@ public void testApi_RemoveInspectionPort_InspectionPortDisappears() throws Excep
     }
+
+    @Test
+    public void testApiRegisterPort() throws Exception {
```
please update the test method names

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/SampleSdnRedirectionApiTest.java:
```diff
@@ -523,4 +524,214 @@ public void testApi_RemoveInspectionPort_InspectionPortDisappears() throws Excep
+    @Test
+    public void testApiUpdateInspectionPort() throws Exception {
+        //Arrange
+        final ArrayList<String> newIngressMacAdresses = new ArrayList<String>(
```
Class variables?

sudhirappaji replied:
removed final and retained as a local variables in the method.

larkinscarvalho commented on src/test/java/org/osc/controller/nsc/SampleSdnRedirectionApiTest.java:
```diff
@@ -523,4 +524,214 @@ public void testApi_RemoveInspectionPort_InspectionPortDisappears() throws Excep
+    @Test
+    public void testApiUpdatePort() throws Exception {
+        //Arrange
+        final ArrayList<String> macAdresses = new ArrayList<String>(
```
you can declare all final variable as class instances

please address multiple places

sudhirappaji replied:
removed final and retained as local variables in the method.

sudhirappaji committed:
Addressed the review comments on Jersey unit tests.

sudhirappaji commented:
@larkinscarvalho , please review and once you approve I will merge these changes.

sudhirappaji commented:
@larkinscarvalho , I am go ahead and merging this... and if you have further remarks, will resolve this as a separate PR.
